database/upload.go

```diff
@@ -52,48 +61,84 @@ func (db *DB) UploadCreate(ctx context.Context, user User, skylink Skylink) (*Up
 }
 
 // UploadsBySkylink fetches all uploads of this skylink
-func (db *DB) UploadsBySkylink(ctx context.Context, skylink Skylink, offset, limit int) ([]Upload, error) {
+func (db *DB) UploadsBySkylink(ctx context.Context, skylink Skylink, offset, limit int) ([]UploadResponseDTO, error) {
 	if skylink.ID.IsZero() {
 		return nil, errors.New("invalid skylink")
 	}
-	filter := bson.D{{"skylink_id", skylink.ID}}
-	opts := options.FindOptions{}
-	if offset > 0 {
-		opts.SetSkip(int64(offset))
-	}
-	if limit > 0 {
-		opts.SetLimit(int64(limit))
-	}
-	c, err := db.staticUploads.Find(ctx, filter, &opts)
-	if err != nil {
-		return nil, err
-	}
-	uploads := make([]Upload, 0)
-	err = c.All(ctx, &uploads)
-	if err != nil {
-		return nil, err
-	}
-	return uploads, nil
+	matchStage := bson.D{{"$match", bson.D{{"skylink_id", skylink.ID}}}}
+	return db.uploadsBy(ctx, matchStage, offset, limit)
 }
 
 // UploadsByUser fetches all uploads by this user
-func (db *DB) UploadsByUser(ctx context.Context, user User, offset, limit int) ([]Upload, error) {
+func (db *DB) UploadsByUser(ctx context.Context, user User, offset, limit int) ([]UploadResponseDTO, error) {
 	if user.ID.IsZero() {
 		return nil, errors.New("invalid user")
 	}
-	filter := bson.D{{"user_id", user.ID}}
-	opts := options.FindOptions{}
-	if offset > 0 {
-		opts.SetSkip(int64(offset))
+	matchStage := bson.D{{"$match", bson.D{{"user_id", user.ID}}}}
+	return db.uploadsBy(ctx, matchStage, offset, limit)
+}
+
+// uploadsBy is a helper function that allows us to fetch a list of downloads,
+// filtered by an arbitrary match criteria.
+//
+// The Mongo query this method executes is
+//	db.uploads.aggregate([
+//		{ $match: { "user_id": ObjectId("5fda32ef6e0aba5d16c0d550") }},
+//		{ $skip: 1 },
+//		{ $limit: 5 },
+//		{ $lookup: {
+//				from: "skylinks",
+//				localField: "skylink_id",  // field in the uploads collection
+//				foreignField: "_id",	   // field in the skylinks collection
+//				as: "fromSkylinks"
+//		  }
+//		},
+//		{ $replaceRoot: { newRoot: { $mergeObjects: [ { $arrayElemAt: [ "$fromSkylinks", 0 ] }, "$$ROOT" ] } } },
+//		{ $project: { fromSkylinks: 0 } },
+//		{ $addFields: { string_id: { $toString: "$_id" } } }
+//	])
+//
+// This query will get all uploads by the current user, skip $skip of them
+// and then fetch $limit of them, allowing us to paginate. It will then
+// join with the `skylinks` collection in order to fetch some additional
+// data about each upload. The last line converts the [12]byte `_id` to hex,
+// so we can easily handle it in JSON.
+func (db *DB) uploadsBy(ctx context.Context, matchStage bson.D, offset, limit int) ([]UploadResponseDTO, error) {
```

Comment:
This code is completely identical to `downloadsBy` except for the return param type and the collection on which we call the aggregation pipeline right? When dealing with aggregation pipelines I usually devise helper functions that help construct the pipeline itself, rather than do the entire thing. That way you end up with things like:

```
const pipeline = getUploadsPipeline(inputA, inputB)
const uploadsDB = await this.mongo.getCollection('uploadsDB') 
return await uploadsDB.aggregate(pipeline)
```

🤷‍♂️  might clean up the code a bit I don't know

Reply:
these helpers look very complex, with the large doc blocks and comments and stuff, but it's "just a list function"

Reply:
I was thinking about doing that but it seemed like it could bite me in the future if I wanted to do different things with the uploads and downloads. Well, YAGNI, I tried your helper approach and the code looks much cleaner. Let's go with that until we need to separate them.